Adds a ticket device and updates brigtimers + QoL

Fruerlund opened the pull request:
Disclaimer! This is the first major duct tape code project of mine. So it's properly filled with stuff that could be done better etc. Just let me know and I will try to improve. Disclaimer!

**Brigdoors**
Brigdoors UI have been reworked and combined with the cell logs. You will now have to input name of the detainee and add/remove charges with the buttons provided. You can choose to pick between existing crimes which will also load the standard time or add custom articles with custom times.
You can still select a whole third time with the Set button or the +/- buttons, both before brigging and after brigging.

![example](https://i.gyazo.com/9c856e8b8d55726676b516263626ad55.png)
![example](https://i.gyazo.com/e9f170769d7c6f86891e32349d0ef66a.png)

***Ticket Device***
This PR adds a new handheld ticket terminal for officers. It can be used to punish for minor crimes by issuing fines. You can choose between the existing minor crimes, or add custom ones with custom ticket amounts. Swipe the ID of the person being ticketed and load their details.

No money, means it will fail to prevent people comitting lots of crimes and just having their account go into minus.

![example](https://i.gyazo.com/8683c22afb346dca7e4ce5d9d85f149f.png)

***Security Records***
Lastly, security computers now have menus for both issued tickets and issued cell logs which can be printed out again if needed. 

🆑Fruerlund
tweak: Updates cell timers
add: a ticketing device for security officers
tweak: adds two new screens to the records computer
/🆑

pinatacolada commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+/obj/item/device/mobileticketplatform
```
can't this name be something shorter?

Fruerlund replied:
The names are not final. Im open for any sugges3

Fruerlund replied:
Suggestions

SamHPurp replied:
simply `ticket_machine`?

SamHPurp commented on code/datums/spacelaw/major.dm:
```diff
@@ -0,0 +1,65 @@
+datum/spacelaw/major/contraband
+	name = "Article 300 Possession of Contraband"
+	desc = "To be in the possession of contraband items. Being in the possession of two or more contraband items, or any restricted Syndicate gadget or weapon, makes you an Enemy of the Corporation (see above). In addition, being caught committing a Major Crime or higher with any of these items makes you an Enemy of the Corporation as well. Personnel turning in these items to Security of their own v ...
```
can you please `\` this only multiple lines, so we don't have to scroll so far over?

SamHPurp commented on code/datums/spacelaw/major.dm:
```diff
@@ -0,0 +1,65 @@
+
+datum/spacelaw/major/assaultofficer
+	name = "Article 302 Assault of an Officer"
+	desc = "To use excessive physical force against a Department Head, Magistrate, NT Rep, or member of Security without the apparent intent to kill them."
```
and here

SamHPurp commented on code/datums/spacelaw/major.dm:
```diff
@@ -0,0 +1,65 @@
+
+datum/spacelaw/major/aggrevatedassault
+	name = "Article 303 Aggrevated Assault"
+	desc = "To assault someone with enough force to cause severe bodily damage (severe external bleeding, severe limb damage, broken bones, internal bleeding, organ damage, disfigurement), but not resulting in their immediate death."
```
and here

SamHPurp commented on code/datums/spacelaw/major.dm:
```diff
@@ -0,0 +1,65 @@
+
+datum/spacelaw/major/majortresspass
+	name = "Article 309 Major Tresspass"
+	desc = "Being in a restricted area without prior authorization. This includes any Security Area, Command area (including EVA), the Engine Room, Atmos, or Toxins Research."
```
and here

Citinited commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+			T.date = current_date_string
+			T.time = worldtime2text()
+			linked_account.transaction_log.Add(T)
+			print_ticket(usr)
```
Shouldn't this be `print_ticket(user)`?

SamHPurp replied:
I can't remember entirely, but I was under the impression we should be using `user` over `usr`... you even are inconsistently using this over this PR.

Fruerlund replied:
I think usr when calling the proc 
and user in the proc

SamHPurp commented on code/datums/spacelaw/major.dm:
```diff
@@ -0,0 +1,65 @@
+
+datum/spacelaw/major/breakingandentryrestricted
+	name = "Article 310 B&E of a Restricted Area"
+	desc = "This is breaking into any Security area, Command area (Bridge, EVA, Captains Quarters, Teleporter, etc), the Engine Room, Atmos, or Toxins research."
```
and here

SamHPurp commented on code/datums/spacelaw/medium.dm:
```diff
@@ -0,0 +1,60 @@
+
+datum/spacelaw/medium/narcoticspossession
+	name = "Article 203 Narcotics Possession"
+	desc = "To distribute narcotics and other controlled substances. This includes ambrosia and space drugs. It is not illegal for them to be grown - distribution is the crime here."
```
and here

SamHPurp commented on code/datums/spacelaw/medium.dm:
```diff
@@ -0,0 +1,60 @@
+
+datum/spacelaw/medium/pettytheft
+	name = "Article 208 Petty Theft"
+	desc = "To take items from areas one does not have authority to access or to take items belonging to others or the station as a whole."
```
and here

SamHPurp commented on code/datums/spacelaw/medium.dm:
```diff
@@ -0,0 +1,60 @@
+
+datum/spacelaw/medium/breakingandentry
+	name = "Article 210 Breaking and Entry"
+	desc = "Forced entry to areas where the subject does not have access to. This counts for general areas, and breaking into restricted areas is a more serious crime."
```
and here

SamHPurp commented on code/datums/spacelaw/minor.dm:
```diff
@@ -0,0 +1,48 @@
+
+datum/spacelaw/minor/drugpossession
+	name = "Article 103 Drug Possession"
+	desc = "To possess space drugs, ambrosia, krokodil, crank, meth, Aranesp, bath salts, THC, or other narcotics, by unauthorized personnel."
```
and here

SamHPurp commented on code/datums/spacelaw/minor.dm:
```diff
@@ -0,0 +1,48 @@
+
+datum/spacelaw/minor/tresspass
+	name = "Article 109 Trespass"
+	desc = "To be in an area which a person does not have access to. This counts for general areas of the ship, and trespass in restricted areas is a more serious crime."
```
and here

SamHPurp commented on code/datums/spacelaw/modifier.dm:
```diff
@@ -0,0 +1,19 @@
+datum/spacelaw/modifiers/surrender
+	name = "Modifier 102 Surrender"
+	desc = "Coming to the brig, confessing what you've done and taking the punishment. Getting arrested without a fuss is not surrender. For this, you have to actually come to the Brig yourself."
```
and here

SamHPurp commented on code/datums/spacelaw/modifier.dm:
```diff
@@ -0,0 +1,19 @@
+
+datum/spacelaw/modifiers/coorporation
+	name = "Modifier 104 Cooperation with prosecution or security"
+	desc = "Being helpful to the members of security, revealing things during questioning or providing names of head revolutionaries."
```
and here

SamHPurp commented on code/datums/spacelaw/modifier.dm:
```diff
@@ -0,0 +1,19 @@
+
+datum/spacelaw/modifiers/refusaltocorporate
+	name = "Modifier 110 Refusal to Cooperate"
+	desc = "When already detained inside the Brig, either in Processing or Temporary Detainment and pending a sentence, refusing to cooperate with Security and instead attempting to break free from their cuffs and/or buckling in such a way that the attending Officers must spend more time preventing their escape than they do actually processing the prisoner."
```
and here

SamHPurp commented on code/datums/spacelaw/modifier.dm:
```diff
@@ -0,0 +1,19 @@
+
+datum/spacelaw/modifiers/repeatoffender
+	name = "Modifier 111 Repeat Offender"
+	desc = "To be repeatedly brigged for the same offense multiple times in a single shift (minimum of three times. Applies on the third brigging). Much like crime stacking, these offenses cannot occur during the same incident. Does not apply to Resisting Arrest."
```
and here

SamHPurp commented on code/datums/spacelaw/major.dm:
```diff
@@ -0,0 +1,65 @@
+datum/spacelaw/major/contraband
+	name = "Article 300 Possession of Contraband"
+	desc = "To be in the possession of contraband items. Being in the possession of two or more contraband items, or any restricted Syndicate gadget or weapon, makes you an Enemy of the Corporation (see above). In addition, being caught committing a Major Crime or higher with any of these items makes you an Enemy of the Corporation as well. Personnel turning in these items to Security of their own v ...
+	min_brig = 10
```
As all `major`/`minor` etc are all the same min/max brig times, it'd be better to just define the parent, then define the `min_brig` `max_brig` times there, rather than having to repeat them absolutely everywhere. ;)

SamHPurp commented on code/game/machinery/computer/security.dm:
```diff
@@ -42,6 +44,18 @@
+		var/obj/item/weapon/paper/P = toprint
+		playsound(loc, "sound/goonstation/machines/printer_dotmatrix.ogg", 50, 1)
+		to_chat(usr, "<span class='notice'>Printing file [P.name].</span>")
+		sleep(50)
```
@tigercat2000 , fox etc... is the plan to be using `sleep` everywhere, or `addtimer` or?..

Fox-McCloud replied:
unless you specifically intend to halt a process, you should be using `addtimer`.

`spawn` is acceptable to be used if you need to spawn of a thread instantly, but sleeping as it's used here? Shouldn't be doing this.

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+	desc = "A device that allows security personnel to issue tickes for violation of Space Law."
+	icon_state = "atmos"
+	item_state = "analyzer"
+	w_class = 2
```
please use the w_class define for this

(from memory: `WEIGHT_CLASS_SMALL`)

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+/obj/item/device/mobileticketplatform
+	name = "handheld ticket dispenser"
+	desc = "A device that allows security personnel to issue tickes for violation of Space Law."
```
pedantry, but `a device that allows FOR security personnel[...]`

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+	w_class = 2
+	slot_flags = SLOT_BELT
+
+	//Variables
```
lol

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+	if(istype(O, /obj/item/weapon/card/id))
+		visible_message("<span class='info'>[user] swipes a card through [src].</span>")
+		if(ID)
+			to_chat(user, "<span class='warning'>Idenfication Card already scanned!</span>")
```
`Idenfication` -> `Identification`

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+		else
+			ID = O
+			D = get_money_account(ID.associated_account_number)
+			to_chat(user, "<span class='warning'>Idenfication Card scanned!</span>")
```
`Idenfication` -> `Identification`

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+			linked_account.transaction_log.Add(T)
+			print_ticket(usr)
+		else
+			to_chat(user, "<span class='warning'>Unable to process requests. User account contains insuffecient funds. Brigging is required.</span>")
```
`insuffecient` -> `insufficient`

as before `/` this onto multiple lines

`Brigging is required.` -> `Incarceration is suggested.`

Last one is more opinion, take/ignore as you like ;)

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+
+
+/obj/item/device/mobileticketplatform/proc/print_ticket(mob/user) // Ticket Reciepts
+		var/ticketid = rand(1111,9999)
```
Might be cool to assign this at the beginning, then increment throughout the round, rather than straight up random each time!

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+	var/datum/money_account/D
+	var/datum/spacelaw/C
+
+	var/list/comittedcrimes = list()
```
`comitted` -> `committed`

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+				clearstringandtime()
+
+	var/datum/spacelaw/selectedcrime = input(user, "Please select article to add for [ID.registered_name]") as null|anything in possible_violations
+	if(!isnull(selectedcrime))
```
erm, unless I'm really misunderstanding the whole `!isnull()` is superfluous, can't you just do `if(selectedcrime)`?

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+/obj/item/device/mobileticketplatform/interact(mob/user as mob)
+	var/dat
+
+	if(screen == 0)
```
Might be nice to use defines for this.

SamHPurp commented on code/game/objects/items/devices/mobileticketplatform.dm:
```diff
@@ -0,0 +1,239 @@
+
+	if(screen == 2)
+		if(!ID)
+			dat = {"<h1>Please swipe ID of the datainee to begin ticketing.</h1><br>
```
`swipe ID` -> `swipe the ID`

`datainee` -> `detainee`

Fruerlund replied:
which one should I use

SamHPurp replied:
the one on the right, for grammatical purposes ;)

Fruerlund replied:
wait, how did I not see that

SamHPurp commented on nano/templates/secure_data.tmpl:
```diff
@@ -112,14 +115,14 @@ Used In File(s): \code\game\machinery\computer\security.dm
 		<h3><center>Menu</center></h3>
 		<div class="line">{{:helper.link('Record Maintenance', 'wrench', {'screen' : 2})}}</div>
-		<div class="line">{{:helper.link('Print Cell Log', 'print', {'printlogs' : 1})}}</div>
+		<div class="line">{{:helper.link('Browse Cell Logs', 'print', {'screen' : 4})}}</div>
```
Looks like you're using the printer icon for browse?

SamHPurp commented on nano/templates/secure_data.tmpl:
```diff
@@ -112,14 +115,14 @@ Used In File(s): \code\game\machinery\computer\security.dm
 		<div class="line">{{:helper.link('Record Maintenance', 'wrench', {'screen' : 2})}}</div>
-		<div class="line">{{:helper.link('Print Cell Log', 'print', {'printlogs' : 1})}}</div>
+		<div class="line">{{:helper.link('Browse Cell Logs', 'print', {'screen' : 4})}}</div>
+		<div class="line">{{:helper.link('Browse Issued Tickets', 'print', {'screen' : 5})}}</div>
```
here too

Fruerlund commented:
Fixed, so MAX brig and MAX fine etc will only be defined once at the top.
Fixed the isnull and the custom articles
Fixed the long ass descriptions \
Fixed typos
Defined screens

I will commit once some more people review this and come up with stuff to fix

SamHPurp commented:
@Fruerlund cool cool, might be best committing anyway so that people can see the whole thing without my marks all over it ;)

Easier to see the bigger picture with your new version

Vivalas commented on code/game/machinery/doors/brigdoors.dm:
```diff
@@ -109,13 +71,130 @@
+	suggestedbrigtime = 0
+	timetoset = 0
+	for(var/datum/spacelaw/C in comittedcrimes)
+		detaineecrimes += "[C.name]<br>"
```
This formats weird on the radio.
![image](https://user-images.githubusercontent.com/6334647/27255638-c163ba22-5367-11e7-8b5d-6c31e3fd8cec.png)

Vivalas replied:
Along with that other thing I mentioned.

Fruerlund commented on code/game/machinery/doors/brigdoors.dm:
```diff
@@ -109,13 +71,130 @@
+
+/obj/machinery/door_timer/proc/update_record(mob/user)
+	var/datum/data/record/R = find_security_record("name", detaineename)
+	Radio.autosay("Detainee: [detaineename] has been incarcerated for; [timetoset/10] seconds for the charges of; [detaineecrimes]\
```
I am looking at it again, I don't think it was my intention to do that. :P

Vivalas replied:
Just a subjective thing, not a big deal.

Fox-McCloud commented:
We generally avoid hard-coding server rules into a feature; this does exactly that. Space law is constantly subject to update and change and is enforced as a server rule rather than a game mechanic.

Other servers/downstreams running paracode shouldn't be obligated to follow our set of rules when utilizing our codebase (not to mention our rules can and do change).

Fox-McCloud commented on code/game/machinery/computer/security.dm:
```diff
@@ -42,6 +44,18 @@
 	ui_interact(user)
 
+/obj/machinery/computer/secure_data/proc/createlog(toprint)
+	if(isnull(toprint))
```
`if(!toprint)`

Fruerlund commented:
@Fox-McCloud I am going to remove the crime datums. I was going to cut it down to one datum for each, and then you'd be able to select the law broken from a list in the datum. Is this something I should do, or should I just remove the datums and stick with same system where you just input crimes and set a time?

outbools commented:
@Fox-McCloud despite the fact there is only one downstream running long enough to worth mention, this is no hard task to modify said datums to fit different set of ingame laws. You can even set a bot to update laws. What about downstreams? We already running into shitton of conflicts and issues yet we are fine at resolving them.

Cash-based cargo is a good idea, i bet you read a lot of those /vg/ economy docs, so you will understand simple issue with cash: there is a lot of places where you can get money yet there is only a few where you can spend them. Being able to get those money from crew via sec tickets sound great, but i think we need to look at how real cop earning their money. You can set a goal each wage cycle to issue x number of tickets and based on this goal progression sec wage can be calculated.

@Fruerlund save current branch for our fork, do not delete it even if it got rejected. We'll have a look at it.

Vivalas commented:
I agree with @outbools. Giving the bureaucracy something to do is not a bad thing. Some friction between bad officers issuing tickets for dosh is a good RP thing and nothing IA can't fix by yelling at command to reject tickets and demoting officers. 

As long as I can remember on almost any codebase I've contributed to the downstream has always been of the least priority. They can either alter or reject the changes or deal with it.

Vivalas commented:
Honestly the crimes datum is what makes this for me, really cool.

taukausanake commented:
So security is now the PoPo. All we need now are mass space pods and a few more secpods around the station and this is perfect

Fruerlund commented:
***Update v1***
Minor bugfixes.
Datums have been reduced since they were just a bunch of copy paste with different names and descriptions.
Everything else as original. You can still choose between existing crimes and custom ones just fine.

Also for: @Vivalas 
![example](https://gyazo.com/3c7b24569c3d13dc3eddb2a70235ee97)

SamHPurp commented on code/game/machinery/doors/brigdoors.dm:
```diff
@@ -19,22 +19,27 @@
+
 	var/list/obj/machinery/targets = list()
-	var/datum/data/record/prisoner
+	var/list/comittedcrimes = list()
```
`committed`

😱

Fruerlund replied:
💯

SamHPurp commented on code/game/machinery/doors/brigdoors.dm:
```diff
@@ -131,47 +136,38 @@
+			selecteddatum = new /datum/spacelaw/
+			selecteddatum.name = input(user, "Please select custom article to add for [detaineename]") as null|text
+			if(selecteddatum.name)
+				selecteddatum.max_brig = input(user, "Please select an amount to add to the brigtime of [detaineename]. The current brigtime is [suggestedbrigtime].") as num
```
suggest using something other than `brigtime`

tigercat2000 commented:
After discussion between the Heads of Staff and Maintainers, it was decided to close this PR. In addition to the issues present with hardcoding something like Space law into our codebase, the system itself was seen as flawed, in that money, in its current form, is completely worthless and devoid of any real value.
